File: searching/dynamic_programming/cpp/combinationSumII.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
    sort(candidates.begin(), candidates.end());  // Sort to handle duplicates easily
    vector<vector<vector<int>>> dp(target + 1);  // DP table

    dp[0] = {{}};

    for (int candidate : candidates) {
        for (int t = target; t >= candidate; --t) {
            for (auto& combination : dp[t - candidate]) {
                combination.push_back(candidate);
                if (find(dp[t].begin(), dp[t].end(), combination) == dp[t].end()) {
                    dp[t].push_back(combination);
                }
                combination.pop_back();
            }
        }
    }

    return dp[target];
}


void printArray(vector<vector<int>> &arr) {
    for (vector<int> & i : arr) {
        for (int &j: i) {
                printf("%d  ", j);
        }
        printf("\n");
    }
}



int main(){
    vector<int> dp = {{10,4,2,7,6,2,5 }};
    vector<vector<int>> value = combinationSum2(dp, 8);
    printArray(value);
    cout << "done" ;
}

File: searching/dynamic_programming/python/combinationSumII.py
def combinationSum2(candidates, target):
    candidates.sort()  # Sort to handle duplicates easily
    dp = [[] for _ in range(target + 1)]
    dp[0] = [[]]  # Base case: one way to make 0, which is the empty set

    for candidate in candidates:
        # Traverse backwards to prevent using the same element multiple times
        for t in range(target, candidate - 1, -1):
            for combination in dp[t - candidate]:
                new_combination = combination + [candidate]
                if new_combination not in dp[t]:
                    dp[t].append(new_combination)

    return dp[target]


# Example usage:
candidates = [10,4,2,7,6,2,5]
target = 8
print(combinationSum2(candidates, target))
